src/lock.rs

```diff
@@ -0,0 +1,328 @@
+    }
+}
+
+impl<'de> Deserialize<'de> for Digest {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_str(DigestVisitor)
+    }
+}
+
+impl PartialEq for Digest {
+    fn eq(&self, other: &Self) -> bool {
+        self.algorithm == other.algorithm && self.bytes == other.bytes
+    }
+}
+
+impl Eq for Digest {}
+
+impl Ord for Digest {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.bytes.cmp(&other.bytes)
+    }
+}
+
+impl PartialOrd for Digest {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// Captures immutable metadata about a given package
+///
+/// It is used to ensure that future installations will use the exact same dependencies.
+#[derive(Serialize, Deserialize, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct LockedPackage {
+    pub name: PackageName,
+    pub digest: Digest,
+    pub registry: RegistryUri,
+    pub repository: String,
+    pub version: Version,
+    pub dependencies: Vec<PackageName>,
+    pub dependants: usize,
+}
+
+impl LockedPackage {
+    /// Captures the source, version and checksum of a Package for use in reproducible installs
+    pub fn lock(
+        package: &Package,
+        registry: RegistryUri,
+        repository: String,
+        dependants: usize,
+    ) -> Self {
+        let digest = digest::digest(&digest::SHA256, &package.tgz)
+            .try_into()
+            .unwrap(); // won't fail as SHA256 is sure to be supported
+
+        Self {
+            name: package.name().to_owned(),
+            registry,
+            repository,
+            digest,
+            version: package.version().to_owned(),
+            dependencies: package
+                .manifest
+                .dependencies
+                .iter()
+                .cloned()
+                .map(|d| d.package)
+                .collect(),
+            dependants,
+        }
+    }
+
+    /// Validates if another LockedPackage matches this one
+    pub fn validate(&self, package: &Package) -> eyre::Result<()> {
+        let digest: Digest = digest::digest(&digest::SHA256, &package.tgz)
+            .try_into()
+            .unwrap();
+
+        ensure!(
+            &self.name == package.name(),
+            "Package name mismatch - expected {}, actual {}",
+            self.name,
+            package.name()
+        );
+        ensure!(
+            self.digest == digest,
+            "Digest mismatch - expected {}, actual {}",
+            self.digest,
+            digest
+        );
+        ensure!(
+            &self.version == package.version(),
+            "Version mismatch - expected {}, actual {}",
+            self.version,
+            package.version()
+        );
+
+        Ok(())
+    }
+
+    /// Constructs a Dependency instance with matching metadata
+    pub fn as_dependency(&self) -> Dependency {
```

Comment:
I think you can only implement this while pinned versions are the only dependency type supported. If you ever anticipate to support more, I would avoid including this `fn` since it _will_ go away in the future.

Reply:
A `Dependency` holds a version requirement which can also be a range – im pretty sure this works in the future

Reply:
The point is that the `Dependency` type is used to fetch the dependency from the registry – if we internally lock the version -> map it as a pinned version im pretty sure it is ok, as this is _the_ job of the lockfile

Reply:
I don't think so, it's always valid to produce a `Dependency` from a` LockedPackage`, since a `LockedPackage` will always have a pinned version. It's the reverse path that isn't valid.

(that said I don't like the API we use for `Registry::download()` expecting a `Dependency`, feels like it should be some more general representation of a package locator — but that's a problem for another PR).